src/anemoi/models/layers/attention.py

```diff
@@ -89,24 +139,220 @@ def forward(
+        # could add something like
+        #   if query.shape[2] != prev_seq_len:
+        #       self.is_attn_compiled = False
+        # To trigger a recompilation
+        if not self.is_attn_compiled:
+            import functools
+
+            from torch.nn.attention.flex_attention import create_block_mask  # should this be after the version check?
+            from torch.nn.attention.flex_attention import flex_attention
+
+            def sliding_window_mask(b, h, q_idx, kv_idx):
+                return abs(q_idx - kv_idx) <= window_size
+
+            seq_len = query.shape[2]
+            self.block_mask = create_block_mask(
+                sliding_window_mask, B=None, H=None, Q_LEN=seq_len, KV_LEN=seq_len, _compile=True
+            )
+            self.attention = functools.partial(
+                flex_attention, block_mask=self.block_mask
+            )  # Cache the block mask (recomended in attn blog post)
+            self.attention = torch.compile(self.attention)
+            self.is_attn_compiled = True
+
+        # TODO test how this impacts scaling at large model counts
+        torch._dynamo.config.optimize_ddp = False
+        out = self.attention(query, key, value)
+        torch._dynamo.config.optimize_ddp = True
 
         return out
+
+
+class FlashAttentionWrapper(nn.Module):
+    """Wrapper for Flash attention."""
+
+    def __init__(self):
+        super().__init__()
+        import flash_attn
+
+        if version.parse(flash_attn.__version__) < version.parse("2.6.0"):
+            raise SystemExit("Error: Flash-attn version is too low. Update to 2.6.0 or higher.")
+        else:
+            self.attention = flash_attn.flash_attn_func
+
+    def forward(
+        self,
+        query,
+        key,
+        value,
+        batch_size: int,
+        causal: bool = False,
+        window_size: int = None,
+        dropout_p: float = 0.0,
+        softcap: float = None,
+        alibi_slopes: torch.Tensor = None,
+    ):
+        query, key, value = (
+            einops.rearrange(t, "batch heads grid vars -> batch grid heads vars") for t in (query, key, value)
+        )
+
+        alibi_slopes = alibi_slopes.repeat(batch_size, 1).to(query.device) if alibi_slopes is not None else None
+
+        out = self.attention(
+            query,
+            key,
+            value,
+            causal=False,
+            window_size=(window_size, window_size),
+            dropout_p=dropout_p,
+            softcap=softcap,
+            alibi_slopes=alibi_slopes,
+        )
+        out = einops.rearrange(out, "batch grid heads vars -> batch heads grid vars")
+        return out
+
+
+def get_alibi_slopes(num_heads: int) -> Tensor:
+    """Calculates linearly decreasing slopes for alibi attention.
+
+    Parameters
+    ----------
+    num_heads : int
+        number of attention heads
+
+    Returns
+    -------
+    Tensor
+        aLiBi slopes
+    """
+    n = 2 ** math.floor(math.log2(num_heads))
```

Comment:
Since `num_heads` is an integer, we could be using bit-shifting here:
`n = 1 << (num_heads.bit_length() - 1)`

Not sure how necessary speed is here though, as a trade-off against readability. It would definitely need a comment.

Reply:
Speed is not an issue as it is only calculated once.